File: iosApp/iosApp/Modules/MnemonicUpdate/MnemonicUpdateWireframe.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

final class DefaultMnemonicUpdateWireframe {
    private weak var parent: UIViewController?
    private let context: MnemonicUpdateWireframeContext
    private let keyStoreService: KeyStoreService
    private let authenticateWireframeFactory: AuthenticateWireframeFactory
    private let alertWireframeFactory: AlertWireframeFactory
    private let settingsService: SettingsService

    private weak var vc: UIViewController?

    init(
        _ parent: UIViewController?,
        context: MnemonicUpdateWireframeContext,
        keyStoreService: KeyStoreService,
        authenticateWireframeFactory: AuthenticateWireframeFactory,
        alertWireframeFactory: AlertWireframeFactory,
        settingsService: SettingsService
    ) {
        self.parent = parent
        self.context = context
        self.keyStoreService = keyStoreService
        self.authenticateWireframeFactory = authenticateWireframeFactory
        self.alertWireframeFactory = alertWireframeFactory
        self.settingsService = settingsService
    }
}

extension DefaultMnemonicUpdateWireframe {

    func present() {
        let vc = wireUp()
        let presentingTopVc = (parent as? UINavigationController)?.topVc
        if settingsService.isSelected(setting: .init(group: .developerTransitions, action: .developerTransitionsCardFlip)) {
            let presentedTopVc = (vc as? UINavigationController)?.topVc
            let delegate = presentedTopVc as? UIViewControllerTransitioningDelegate
            self.vc = vc
            vc.modalPresentationStyle = .overFullScreen
            vc.transitioningDelegate = delegate
        } else if settingsService.isSelected(setting: .init(group: .developerTransitions, action: .developerTransitionsSheet)) {
            vc.modalPresentationStyle = .automatic
        }
        presentingTopVc?.present(vc, animated: true)
    }

    func navigate(with destination: MnemonicUpdateWireframeDestination) {
        if let input = destination as? MnemonicUpdateWireframeDestination.Authenticate {
            authenticateWireframeFactory.make(vc, context: input.context).present()
        }
        if let input = destination as? MnemonicUpdateWireframeDestination.Alert {
            alertWireframeFactory.make(vc, context: input.context).present()
        }
        if destination is MnemonicUpdateWireframeDestination.Dismiss {
            // NOTE: Dispatching on next run loop so that presenting
            // controller collectionView has time to reload and does not
            // break custom dismiss animation
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.01) {
                [weak self] in self?.vc?.popOrDismiss()
            }
        }
    }
}

private extension DefaultMnemonicUpdateWireframe {

    func wireUp() -> UIViewController {
        let interactor = DefaultMnemonicUpdateInteractor(
            keyStoreService: keyStoreService
        )
        let vc: MnemonicUpdateViewController = UIStoryboard(.mnemonicUpdate).instantiate()
        let presenter = DefaultMnemonicUpdatePresenter(
            view: WeakRef(referred: vc),
            wireframe: self,
            interactor: interactor,
            context: context
        )
        vc.presenter = presenter
        let nc = NavigationController(rootViewController: vc)
        self.vc = nc
        return nc
    }
}

extension DefaultMnemonicUpdateWireframe {

    enum Constant {
        static let saltExplanationURL = URL(
            string: "https://www.youtube.com/watch?v=XqB5xA62gLw"
        )!
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/mnemonicupdate/MnemonicUpdateWireframe.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.mnemonicupdate

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.keyStore.KeyStoreService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.extensions.navigationFragment
import com.sonsofcrypto.web3wallet.android.common.ui.navigationFragment
import com.sonsofcrypto.web3wallet.android.modules.compose.alert.AlertWireframeFactory
import com.sonsofcrypto.web3walletcore.modules.mnemonicUpdate.*

class DefaultMnemonicUpdateWireframe(
    private val parent: WeakRef<Fragment>?,
    private val context: MnemonicUpdateWireframeContext,
    private val alertWireframeFactory: AlertWireframeFactory,
    private val keyStoreService: KeyStoreService,
): MnemonicUpdateWireframe {

    override fun present() {
        val fragment = wireUp()
        parent?.navigationFragment?.present(fragment, animated = true)
    }

    override fun navigate(destination: MnemonicUpdateWireframeDestination) {
        when (destination) {
            is MnemonicUpdateWireframeDestination.Alert -> {
                alertWireframeFactory.make(parent?.get(), destination.context).present()
            }
            is MnemonicUpdateWireframeDestination.Dismiss -> {
                parent?.navigationFragment?.dismiss()
            }
            else -> {
                println("[AAA] Implement -> $destination")
            }
        }
    }

    private fun wireUp(): Fragment {
        val view = MnemonicUpdateFragment()
        val interactor = DefaultMnemonicUpdateInteractor(
            keyStoreService,
        )
        val presenter = DefaultMnemonicUpdatePresenter(
            WeakRef(view),
            this,
            interactor,
            context,
        )
        view.presenter = presenter
        return view
    }
}